From 7cf970208c1766fcf686841df709e79daf0078af Mon Sep 17 00:00:00 2001
From: Patrick Tjin <pattjin@google.com>
Date: Fri, 21 Apr 2017 23:34:24 +0000
Subject: [PATCH 0110/1240] net: wireless: bcmdhd: Do not print out device name
 on invalid length

(cherry picked from commit b149dd5d22c3e4c2faab0bb934a018888ff99ef3)

Signed-off-by: Patrick Tjin <pattjin@google.com>

Bug: 27335848

[fixes CVE-2016-0801]
Signed-off-by: Kees Cook <keescook@chromium.org>
Change-Id: Id5ae94c50f486f2f173def9b616322111881815a
---
 drivers/net/wireless/bcmdhd/wl_cfg80211.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/drivers/net/wireless/bcmdhd/wl_cfg80211.c b/drivers/net/wireless/bcmdhd/wl_cfg80211.c
index d95ba55..7d05839 100644
--- a/drivers/net/wireless/bcmdhd/wl_cfg80211.c
+++ b/drivers/net/wireless/bcmdhd/wl_cfg80211.c
@@ -1325,10 +1325,12 @@ wl_validate_wps_ie(char *wps_ie, s32 wps_ie_len, bool *pbc)
 		} else if (subelt_id == WPS_ID_DEVICE_NAME) {
 			char devname[100];
 			size_t namelen = MIN(subelt_len, sizeof(devname));
-			memcpy(devname, subel, namelen);
-			devname[namelen-1] = '\0';
-			WL_DBG(("  attr WPS_ID_DEVICE_NAME: %s (len %u)\n",
-				devname, subelt_len));
+			if (namelen) {
+				memcpy(devname, subel, namelen);
+				devname[namelen - 1] = '\0';
+				WL_DBG(("  attr WPS_ID_DEVICE_NAME: %s (len %u)\n",
+					devname, subelt_len));
+			}
 		} else if (subelt_id == WPS_ID_DEVICE_PWD_ID) {
 			valptr[0] = *subel;
 			valptr[1] = *(subel + 1);
-- 
2.7.4

